### PR TITLE
[Flax] Add wav2vec2

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Adds Wav2Vec2 in Flax

- [x] FlaxCTCWav2Vec2
- [x] FlaxForPreTraining
- [x] FlaxWav2Vec2 random mask code
- [x] Clean-up
- [x] Write pretraining script

### CODE REVIEW

**Comment by ThomAub** [comment]
Anything I can help with ? 
Looking forward to use Flasx Wav2vec for the community event :) :)

**Comment by patrickvonplaten** [comment]
Hey @ThomAub - that's very nice of you! I will need to spend a couple more hours on this and then add pretraining as well. One thing that would be very helpful would be to check how to do the GumbelSoftmax in Flax. *I.e.* how can we translate this: https://github.com/huggingface/transformers/blob/f2c4ce7e339f4a2f8aaacb392496bc1a5743881f/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L751 PyTorch module to Flax? This might be a bit difficult and require some googling to see if others have already implement gumbel softmax in jax/Flax or not. If you could take a look at this, it would be very useful!

**Comment by ThomAub** [comment]
I guess I wasn't fast enough ! Great work

**Comment by patrickvonplaten** [comment]
@patil-suraj @sgugger - actually this is still WIP. Sorry for tagging you too early

File: examples/research_projects/jax-projects/wav2vec2/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+Pretraining Wav2Vec2 is rather complex, so it is highly recommended to read the 
+[official paper](https://arxiv.org/abs/2006.11477).
+
+JAX/Flax allows you to trace pure functions and compile them into efficient, fused accelerator code on both GPU and TPU.
+Models written in JAX/Flax are **immutable** and updated in a purely functional
+way which enables simple and efficient model parallelism.
+
+`run_wav2vec2_pretrain_flax.py` is a lightweight example of how to download and preprocess a dataset from the 🤗 Datasets library or use your own files (jsonlines or csv), then pretrain the wav2vec2 architectures above on it.
+
+For custom datasets in `jsonlines` format please see: https://huggingface.co/docs/datasets/loading_datasets.html#json-files and you also will find examples of these below.
```
Comment (sgugger) [inline examples/research_projects/jax-projects/wav2vec2/README.md:13]:
```suggestion
For custom datasets in `jsonlines` format please see: [the Datasets documentation](https://huggingface.co/docs/datasets/loading_datasets.html#json-files) and you also will find examples of these below.
```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 2]
```diff
@@ -0,0 +1,1214 @@
...
+
+@flax.struct.dataclass
+class FlaxWav2Vec2BaseModelOutput(ModelOutput):
+    """
+    Output type of :class:`~transformers.FlaxWav2Vec2BaseModelOutput`, with potential hidden states and attentions.
+
+    Args:
+        last_hidden_state (:obj:`jnp.ndarray` of shape :obj:`(batch_size, sequence_length, hidden_size)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        extract_features (:obj:`jnp.ndarray` of shape :obj:`(batch_size, sequence_length, conv_dim[-1])`):
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:48]: conv_dim[-1] is not super informative for someone reading the doc, can we say something clearer?

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 3]
```diff
@@ -0,0 +1,1214 @@
...
+
+
+def _sample_negative_indices(features_shape: Tuple, num_negatives: int):
+    """
+    Sample `num_negatives` vectors from feature vectors.
+    """
+    batch_size, sequence_length, hidden_size = features_shape
+    if sequence_length <= 1:
+        raise ValueError(
+            f"`features should have `sequence_length` > 1, but are of shape (batch_size, sequence_length, hidden_size) = ({batch_size, sequence_length, hidden_size})."
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:178]: Split on two lines please :-)

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+        hidden_states = self.intermediate_dense(hidden_states)
+        hidden_states = self.intermediate_act_fn(hidden_states)
+        hidden_states = self.intermediate_dropout(hidden_states, deterministic=deterministic)
+
+        hidden_states = self.output_dense(hidden_states)
+        hidden_states = self.output_dropout(hidden_states, deterministic=deterministic)
+        return hidden_states
+
+
+class FlaxWav2Vec2EncoderLayerStableLayerNorm(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:532]: Those names are starting to get out of hand :fearful:

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (attn_weights,)
+
+        return outputs
+
+
+class FlaxWav2Vec2EncoderLayerStableLayerNormCollection(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:568]: :fearful: :fearful: :fearful: :fearful:

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+    config: Wav2Vec2Config
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.num_groups = self.config.num_codevector_groups
+        self.num_vars = self.config.num_codevectors_per_group
+
+        if self.config.codevector_dim % self.num_groups != 0:
+            raise ValueError(
+                f"`config.codevector_dim {self.config.codevector_dim} must be divisible by `config.num_codevector_groups` {self.num_groups} for concatenation"
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:680]: Split on several lines please.

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+            perplexity = self._compute_perplexity(codevector_probs, mask_time_indices)
+
+        codevector_probs = codevector_probs.reshape(batch_size * sequence_length, -1)
+        # use probs to retrieve codevectors
+        codevectors_per_group = jnp.expand_dims(codevector_probs, axis=-1) * self.codevectors
+        codevectors = (
+            codevectors_per_group.reshape(batch_size * sequence_length, self.num_groups, self.num_vars, -1)
+            .sum(-2)
+            .reshape(batch_size, sequence_length, -1)
+        )
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:740]: I think it would be clearer to use three lines instead of grouping it all together in one like this.

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+    ):
+        r"""
+        Returns:
+
+        Example::
+
+            >>> import optax
+            >>> import numpy as np
+            >>> import jax.numpy as jnp
+            >>> from transformers import Wav2Vec2FeatureExtractor, Wav2Vec2ForPreTraining
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:1077]:
```suggestion
            >>> from transformers import Wav2Vec2FeatureExtractor, FlaxWav2Vec2ForPreTraining
```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        r"""
+        Returns:
+
+        Example::
+
+            >>> import jax.numpy as jnp
+            >>> from transformers import Wav2Vec2Processor, Wav2Vec2ForCTC
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:970]:
```suggestion
            >>> from transformers import Wav2Vec2Processor, FlaxWav2Vec2ForCTC
```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 10]
```diff
@@ -0,0 +1,1214 @@
...
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        """
+
+        Returns:
+
+        Example::
+
+            >>> from transformers import Wav2Vec2Processor, Wav2Vec2Model
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:854]:
```suggestion
            >>> from transformers import Wav2Vec2Processor, FlaxWav2Vec2Model
```

File: tests/test_modeling_flax_wav2vec2.py [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+            2, 0, 1, 3
+        )
+
+        self.assertTrue(negatives.shape == (num_negatives, batch_size, sequence_length, hidden_size))
+
+        # make sure no negatively sampled vector is actually a positive one
+        for negative in negatives:
+            self.assertTrue(((negative - features.reshape(negative.shape)) == 0).sum() == 0.0)
+
+        # make sure that full vectors are sampled and not values of vectors => this means that `unique()` yields a single value for `hidden_size` dim
```
Comment (sgugger) [inline tests/test_modeling_flax_wav2vec2.py:287]: Split on several lines please :-)

**Review by sgugger — APPROVED** [review]
Great addition, thanks a lot!

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 12]
```diff
@@ -0,0 +1,1214 @@
...
+
+    def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple) -> FrozenDict:
+        # init input tensors
+        input_ids = jnp.zeros(input_shape, dtype="i4")
+        attention_mask = jnp.ones_like(input_ids)
+        params_rng, dropout_rng = jax.random.split(rng, 2)
+        rngs = {"params": params_rng, "dropout": dropout_rng}
+
+        return self.module.init(rngs, input_ids, attention_mask, return_dict=False)["params"]
+
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:774]:
```suggestion
        input_values = jnp.zeros(input_shape, dtype="f4")
        attention_mask = jnp.ones_like(input_ids)
        params_rng, dropout_rng = jax.random.split(rng, 2)
        rngs = {"params": params_rng, "dropout": dropout_rng}

        return self.module.init(rngs, input_values, attention_mask, return_dict=False)["params"]

```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 13]
```diff
@@ -0,0 +1,1214 @@
...
+class FlaxWav2Vec2Module(nn.Module):
+    config: Wav2Vec2Config
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.feature_extractor = FlaxWav2Vec2FeatureExtractor(self.config, dtype=self.dtype)
+        self.feature_projection = FlaxWav2Vec2FeatureProjection(self.config, dtype=self.dtype)
+        self.masked_spec_embed = self.param(
+            "masked_spec_embed", jax.nn.initializers.uniform(), (self.config.hidden_size,)
+        )
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:831]: dtype ?

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 14]
```diff
@@ -0,0 +1,1214 @@
...
+class FlaxWav2Vec2Model(FlaxWav2Vec2PreTrainedModel):
+    module_class = FlaxWav2Vec2Module
+
+
+class FlaxWav2Vec2ForCTCModule(nn.Module):
+    config: Wav2Vec2Config
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.wav2vec2 = FlaxWav2Vec2Module(self.config)
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:946]:
```suggestion
        self.wav2vec2 = FlaxWav2Vec2Module(self.config, dtype=self.dtype)
```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 15]
```diff
@@ -0,0 +1,1214 @@
...
+class FlaxWav2Vec2ForCTC(FlaxWav2Vec2PreTrainedModel):
+    module_class = FlaxWav2Vec2ForCTCModule
+
+
+class FlaxWav2Vec2ForPreTrainingModule(nn.Module):
+    config: Wav2Vec2Config
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.wav2vec2 = FlaxWav2Vec2Module(self.config)
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:1043]:
```suggestion
        self.wav2vec2 = FlaxWav2Vec2Module(self.config, dtype=self.dtype)
```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 16]
```diff
@@ -0,0 +1,1214 @@
...
+        self.dropout_features = nn.Dropout(self.config.feat_quantizer_dropout)
+
+        self.quantizer = FlaxWav2Vec2GumbelVectorQuantizer(self.config)
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:1046]:
```suggestion
        self.quantizer = FlaxWav2Vec2GumbelVectorQuantizer(self.config, dtype=self.dtype)
```

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py [hunk 17]
```diff
@@ -0,0 +1,1214 @@
...
+        """
+
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        outputs = self.wav2vec2(
+            input_values,
+            attention_mask=attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            mask_time_indices=mask_time_indices,
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:1121]:
```suggestion
            mask_time_indices=mask_time_indices,
            deterministic=deterministic,
```

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for adding this so quickly!

Looks good apart from missing a few `dtypes` and a couple of really long names.

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py [hunk 18]
```diff
@@ -0,0 +1,563 @@
...
+        cache_dir=model_args.cache_dir,
+        gradient_checkpointing=model_args.gradient_checkpointing,
+    )
+
+    if not config.do_stable_layer_norm or config.feat_extract_norm != "layer":
+        raise ValueError(
+            "PreTraining is only supported for ``config.do_stable_layer_norm=True`` and ``config.feat_extract_norm='layer'"
+        )
+
+    model = FlaxWav2Vec2ForPreTraining(config)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py:339]: great catch!

File: src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:532]: True since they are not public, I'll change them across all TF, PT, and Flax in a follow-up PR

Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py:532]: https://github.com/huggingface/transformers/issues/12443

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py [hunk 19]
```diff
@@ -0,0 +1,563 @@
...
+            Maximum length of the ``input_values`` of the returned list and optionally padding length (see above).
+        pad_to_multiple_of (:obj:`int`, `optional`):
+            If set will pad the sequence to a multiple of the provided value.
+            This is especially useful to enable the use of Tensor Cores on NVIDIA hardware with compute capability >=
+            7.5 (Volta).
+    """
+
+    model: FlaxWav2Vec2ForPreTraining
+    feature_extractor: Wav2Vec2FeatureExtractor
+    padding: Union[bool, str] = "longest"
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py:150]: think "longest" is fine here since the range differs too much
